File: LEETCODE/LEETCODE-CPP/110.cpp
//110. Balanced Binary Tree
//https://leetcode.com/problems/balanced-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int leftHeight=0, rightHeight=0;
        
        if(root==NULL)
            return true;

        leftHeight=helper(root->left);
        rightHeight=helper(root->right);

        bool flag=abs(leftHeight-rightHeight)<=1;
        
        bool leftAns= isBalanced(root->left);
        bool rightAns= isBalanced(root->right);
        return flag&&leftAns&&rightAns;        
    }

    int helper(TreeNode* root){
        if(root==NULL)
            return 0;

        int lh = helper(root->left);
        int rh = helper(root->right);

        int maxHeight = max(lh, rh);
        
        return maxHeight + 1;
    }
};

File: LEETCODE/LEETCODE-PYTHON/110.py
# 110. Balanced Binary Tree
# https://leetcode.com/problems/balanced-binary-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True

        def helper(root):
            if root is None:
                return -1

            return max(helper(root.left),helper(root.right))+1

        return (
            abs(helper(root.left) - helper(root.right)) <2
            and self.isBalanced(root.left)
            and self.isBalanced(root.right)
        )